### PR TITLE
Extend the GPIO HAL to support a port/pin numbering scheme

### PR DESCRIPTION
**Pull request by g-oikonomou**
This pull changes the GPIO HAL such that it supports a port/pin numbering scheme for pins.

What I am proposing is having two modes, controlled through `GPIO_HAL_PORT_PIN_NUMBERING`:

* 0: Selects what we have currently, whereby pins are numbered sequentially from 0, up to possibly 63. Multiple pins can be manipulated in a single go using a 32-bit or 64-bit mask (mask size dependent on pin count).
* 1 (default): Selects a port/pin numbering. A port is represented by a `uint8_t` and each port can have up to 32 pins, each pin also represented by a `uint8_t`. Multiple pins _within the same port_ can still be manipulated in a single go using a 32-bit mask.

Depending on choice of mode, functions come in two flavours, `port_pin_` and `no_port_pin_`. For example:

```
void gpio_hal_arch_port_pin_set_input(gpio_hal_port_t port, gpio_hal_pin_t pin);
void gpio_hal_arch_no_port_pin_set_input(gpio_hal_pin_t pin);
```

Macros are used to automatically map to the correct one, eg
```
#if GPIO_HAL_PORT_PIN_NUMBERING
#define gpio_hal_arch_pin_set_input(port, pin) \
  gpio_hal_arch_port_pin_set_input(port, pin)
#else
#define gpio_hal_arch_pin_set_input(port, pin) \
  gpio_hal_arch_no_port_pin_set_input(pin)
#endif /* GPIO_HAL_PORT_PIN_NUMBERING */
```

I need to double check a couple of things and also this will need a rebase after #671, #683, #685, so I am flagging this as WIP for now.

Your comments please, @joakimeriksson hopefully this meets your requirements for your new port.

### CODE REVIEW

**Comment by joakimeriksson** [comment]
@g-oikonomou  Yes, this would work for me on the EFR32. I will implement it and test it on the EFR - but from what I can see it is good!

**Comment by joakimeriksson** [comment]
I have now tested this fully on EFR32 and this API is fine for me. Maybe in the long run we should go for port, pin style everywhere to avoid so many pre-processor ifs. (and make use of the NULLPORT if it is not used on a specific platform).

**Comment by g-oikonomou** [comment]
OK great!

This needs some conflict fixes and also I need to have a quick look at the simplelink port to make sure it's OK

**Comment by joakimeriksson** [comment]
Yes, but if we pull the EFR32 PR this will get in also. So not sure what is best?

**Comment by g-oikonomou** [comment]
Happy to fix conflicts and merge separately as well as to merge at the same time as the EFR32 port. Depends how quickly the EFR32 can be ready to merge :)

**Comment by g-oikonomou** [comment]
So, this is now all nicely rebased.

I have triple-checked against #571 and there are two differences:

* #571 adds `gpio_hal_arch_init(void)`. I suggest doing this as part of #571 since this is where it's needed
* This pull has an additional commit (768326002), but pulling it into #571 is trivial.

As discussed earlier, these changes here address previous concerns about the API for platforms with a large number of pins and there is no reason to delay this until the EFR32 port is ready.

In other words, please review and merge ;)

Fixes #538
